### PR TITLE
Trusted entitlements: Support post params hash header

### PR DESCRIPTION
**Pull request by tonidero**
### Description
As an additional security measure we will add a new header to the request that will be part of the data to verify as part of the signature.

Depends on https://github.com/RevenueCat/khepri/pull/6246

### CODE REVIEW

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt [hunk 1]
```diff
@@ -326,13 +339,19 @@ internal class Backend(
...
         val call = object : Dispatcher.AsyncCall() {
             override fun call(): HTTPResult {
+                val body = mapOf(
+                    APP_USER_ID to appUserID,
+                    NEW_APP_USER_ID to newAppUserID,
+                )
+                val postFieldsToSign = listOf(
+                    APP_USER_ID to appUserID,
+                    NEW_APP_USER_ID to newAppUserID,
+                )
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt:349]: I thought about just copying the `body` (modifying it to be a list of key-values) but in the end, I thought it's better to keep it as a separate entity.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/HTTPClient.kt [hunk 2]
```diff
@@ -117,23 +119,24 @@ internal class HTTPClient(
...
         if (callResult == null) {
             log(LogIntent.WARNING, NetworkStrings.ETAG_RETRYING_CALL)
-            callResult = performRequest(baseURL, endpoint, body, requestHeaders, refreshETag = true)
+            callResult = performRequest(baseURL, endpoint, body, postFieldsToSign, requestHeaders, refreshETag = true)
         }
         return callResult
     }
 
-    @Suppress("ThrowsCount")
+    @Suppress("ThrowsCount", "LongParameterList", "LongMethod")
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/HTTPClient.kt:134]: I want to refactor these functions to not be so long or require so many parameters... But that should be done in a different PR.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/verification/SigningManager.kt [hunk 3]
```diff
@@ -75,7 +81,33 @@ internal class SigningManager(
     fun createRandomNonce(): String {
         val bytes = ByteArray(NONCE_BYTES_SIZE)
         SecureRandom().nextBytes(bytes)
-        return String(Base64.encode(bytes, Base64.DEFAULT))
+        return String(Base64.encode(bytes, Base64.DEFAULT)).trim()
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/verification/SigningManager.kt:84]: This `trim` was needed for the backend integration tests that were added to work... Basically we were always creating nonces with a final `\n` character. It worked on Android fine it seems since the HttpUrlConnection implementation works different than the one in unit tests. But it was complaining in the tests.

I've confirmed that this doesn't break anything in our test app and with our integration tests.

File: purchases/src/test/java/com/revenuecat/purchases/common/verification/SigningManagerTest.kt [hunk 4]
```diff
@@ -269,8 +333,46 @@ class SigningManagerTest {
...
         requestTime: String? = "1677005916012",
-        eTag: String? = null
-    ) = signingManager.verifyResponse(requestPath, signature, nonce, body, requestTime, eTag)
+        eTag: String? = null,
+        postParamsHeader: String? = null,
+    ) = signingManager.verifyResponse(requestPath, signature, nonce, body, requestTime, eTag, postParamsHeader)
+
+    // We can use this function to create new signatures if we need to change the test data
+    @Suppress("Unused")
+    private fun createFakeSignature(
```
Comment (tonidero) [inline purchases/src/test/java/com/revenuecat/purchases/common/verification/SigningManagerTest.kt:342]: This is a messy implementation to create signatures for the tests. I didn't check it in before, but I think it's useful. I don't think we should use it during the tests though since it has a lot of logic. I prefer to hardcode the expected signature.

**Comment by tonidero** [comment]
Backend integration tests are failing since they depend on https://github.com/RevenueCat/khepri/pull/6246

**Comment by tonidero** [comment]
Holding until backend is deployed.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/verification/SigningManager.kt (hunk 3 above)
Comment (NachoSoto) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/verification/SigningManager.kt:84]: It might help to write a test for this?

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/verification/SigningManager.kt [hunk 5]
```diff
@@ -75,7 +81,33 @@ internal class SigningManager(
...
+        endpoint: Endpoint,
+        postFieldsToSign: List<Pair<String, String>>?,
+    ): String? {
+        return if (!postFieldsToSign.isNullOrEmpty() &&
+            shouldVerifyEndpoint(endpoint)
+        ) {
+            val sha256Digest = MessageDigest.getInstance("SHA-256")
+            postFieldsToSign.mapIndexed { index, pair ->
+                if (index > 0) {
+                    sha256Digest.update(0x00.toByte())
```
Comment (NachoSoto) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/verification/SigningManager.kt:97]: Maybe extract the separator to a constant?

File: purchases/src/test/java/com/revenuecat/purchases/common/verification/SigningManagerTest.kt [hunk 6]
```diff
@@ -96,6 +105,44 @@ class SigningManagerTest {
...
+    @Test
+    fun `getPostParamsForSigningHeaderIfNeeded returns header with correct value in informational`() {
+        val header = informationalSigningManager.getPostParamsForSigningHeaderIfNeeded(Endpoint.LogIn, testFieldsToSign)
+        assertThat(header).isEqualTo("test-field-1,test-field-2:sha256:0a1d806f908079361fc0d18073dfcfcacb730afe4ace3d0478602479892e81d3")
+    }
+
+    @Test
+    fun `getPostParamsForSigningHeaderIfNeeded returns header with correct value in enforced`() {
+        val header = enforcedSigningManager.getPostParamsForSigningHeaderIfNeeded(Endpoint.LogIn, testFieldsToSign)
+        assertThat(header).isEqualTo("test-field-1,test-field-2:sha256:0a1d806f908079361fc0d18073dfcfcacb730afe4ace3d0478602479892e81d3")
```
Comment (NachoSoto) [inline purchases/src/test/java/com/revenuecat/purchases/common/verification/SigningManagerTest.kt:143]: 👍🏻
